types/src/contract_wasm.rs

```diff
@@ -1,11 +1,226 @@
-use core::fmt::Debug;
+use alloc::{format, string::String, vec::Vec};
+use core::{
+    array::TryFromSliceError,
+    convert::TryFrom,
+    fmt::{self, Debug, Display, Formatter},
+};
 
-use serde::Serialize;
+use datasize::DataSize;
+#[cfg(feature = "std")]
+use schemars::{gen::SchemaGenerator, schema::Schema, JsonSchema};
+use serde::{de::Error as SerdeError, Deserialize, Deserializer, Serialize, Serializer};
 
-use crate::bytesrepr::{Bytes, Error, FromBytes, ToBytes};
+use crate::{
+    account,
+    account::TryFromSliceForAccountHashError,
+    bytesrepr::{Bytes, Error, FromBytes, ToBytes},
+    uref, CLType, CLTyped, HashAddr,
+};
 
 const CONTRACT_WASM_MAX_DISPLAY_LEN: usize = 16;
+const KEY_HASH_LENGTH: usize = 32;
+const WASM_STRING_PREFIX: &str = "contract-wasm-";
+
+/// Associated error type of `TryFrom<&[u8]>` for `ContractHash`.
+#[derive(Debug)]
+pub struct TryFromSliceForContractHashError(());
+
+#[derive(Debug)]
+pub enum FromStrError {
+    InvalidPrefix,
+    Hex(base16::DecodeError),
+    Account(TryFromSliceForAccountHashError),
+    Hash(TryFromSliceError),
+    AccountHash(account::FromStrError),
+    URef(uref::FromStrError),
+}
+
+impl From<base16::DecodeError> for FromStrError {
+    fn from(error: base16::DecodeError) -> Self {
+        FromStrError::Hex(error)
+    }
+}
+
+impl From<TryFromSliceForAccountHashError> for FromStrError {
+    fn from(error: TryFromSliceForAccountHashError) -> Self {
+        FromStrError::Account(error)
+    }
+}
+
+impl From<TryFromSliceError> for FromStrError {
+    fn from(error: TryFromSliceError) -> Self {
+        FromStrError::Hash(error)
+    }
+}
+
+impl From<account::FromStrError> for FromStrError {
+    fn from(error: account::FromStrError) -> Self {
+        FromStrError::AccountHash(error)
+    }
+}
+
+impl From<uref::FromStrError> for FromStrError {
+    fn from(error: uref::FromStrError) -> Self {
+        FromStrError::URef(error)
+    }
+}
+
+impl Display for FromStrError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            FromStrError::InvalidPrefix => write!(f, "invalid prefix"),
+            FromStrError::Hex(error) => write!(f, "decode from hex: {}", error),
+            FromStrError::Account(error) => write!(f, "account from string error: {:?}", error),
+            FromStrError::Hash(error) => write!(f, "hash from string error: {}", error),
+            FromStrError::AccountHash(error) => {
+                write!(f, "account hash from string error: {:?}", error)
+            }
+            FromStrError::URef(error) => write!(f, "uref from string error: {:?}", error),
+        }
+    }
+}
+
+/// A newtype wrapping a `HashAddr` which is the raw bytes of
+/// the ContractWasmHash
+#[derive(DataSize, Default, PartialOrd, Ord, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct ContractWasmHash(HashAddr);
+
+impl ContractWasmHash {
+    /// Constructs a new `ContractWasmHash` from the raw bytes of the contract hash.
```

Comment:
You forgot to replace "contract hash" with "contract wasm hash" in the comments.

Reply:
Replaced